### ISSUE TITLE
new magestead with-bootstrap-patch error

### ISSUE THREAD

**Issue by veilig2000** [richdynamix/magestead#55]
I just created a new magestead project `magestead new magento.magestead.clean`. After everything finished provisioning, I updated my composer to set the magento composer installer to bootstrap for the composer autoloader. `"with-bootstrap-patch": true`

```
{                                                                                
    "name": "richdynamix/magestead",                                             
    "description": "Magento build for ...",                                      
    "require-dev": {                                                             
        "magetest/magento-phpspec-extension": "3.0.0",                           
        "sensiolabs/behat-page-object-extension": "2.0.*@dev",                   
        "magetest/magento-behat-extension": "dev-feature/Behat3",                
        "magetest/manager": "*",                                                 
        "behat/mink": "1.6.*",                                                   
        "behat/mink-goutte-driver": "1.1.*",                                     
        "behat/mink-selenium2-driver": "*",                                      
        "behat/mink-extension": "*",                                             
        "phpunit/phpunit": "3.7.*",                                              
        "phpunit/dbunit": "1.3.*",                                               
        "phpunit/phpunit-selenium": "1.3.*",                                     
        "phpunit/phpunit-story": "1.0.*",                                        
        "squizlabs/php_codesniffer": "1.*",                                      
        "phpmd/phpmd": "1.*",                                                    
        "sebastian/phpcpd": "2.0.*",                                             
        "pdepend/pdepend": "1.*",                                                
        "phploc/phploc": "2.*",                                                  
        "phpdocumentor/phpdocumentor": "2.*",                                    
        "theseer/phpdox": "0.6.*",                                               
        "theseer/fxsl": "1.0.*@dev",                                             
        "mayflower/php-codebrowser": "~1.1",                                     
        "darsadow/behat-parameters-extension": "dev-master",                     
        "bossa/phpspec2-expect": "*",                                            
        "ciaranmcnulty/behat-stepthroughextension": "dev-master"                 
    },                                                                           
    "require": {                                                                 
        "aydin-hassan/magento-core-composer-installer" : "~1.0",                 
        "magento/magento" : "1.9.2.2",                                           
        "magento-hackathon/magento-composer-installer": "3.0.*",                 
        "inchoo/php7": "1.0.*"                                                   
    },                                                                           
    "authors": [                                                                 
        {                                                                        
            "name": "Steven Richardson",                                         
            "email": "steven@richdynamix.com"                                    
        }                                                                        
    ],                                                                           
    "repositories": [                                                            
        {                                                                        
            "type": "vcs",                                                       
            "url": "git@github.com:AydinHassan/magento-community.git"            
        },                                                                       
        {                                                                        
            "type": "composer",                                                  
            "url": "https://packages.firegento.com"                              
        }                                                                        
    ],                                                                           
    "extra": {                                                                   
        "magento-root-dir": "public",                                            
        "magento-force": true,                                                   
        "auto-append-gitignore": true,                                           
        "with-bootstrap-patch": true                                             
    },                                                                           
    "config": {                                                                  
        "bin-dir": "bin"                                                         
    },                                                                           
    "autoload": {                                                                
        "psr-0": {                                                               
            "MageTest\\PhpSpec\\MagentoExtension": "src",                        
            "": [                                                                
                "public/app/code/local",                                         
                "public/app/code/community",                                     
                "public/app/code/core",                                          
                "public/app",                                                    
                "public/lib",                                                    
                "features/bootstrap"                                             
            ],                                                                   
            "Mage": "public/app/code/core"                                       
        }                                                                        
    }                                                                            
}
```

Afterwards, I redeployed so autoloading patching would take place.

`composer run-script post-install-cmd -vvv -- --redeploy`

```
Reading ./composer.json
Loading config file /Users/veilig/.composer/config.json
Loading config file /Users/veilig/.composer/auth.json
Loading config file ./composer.json
Checked CA file /usr/local/etc/openssl/cert.pem: valid
Executing command (/Users/veilig/Sites/magento.magestead.clean): git branch --no-color --no-abbrev -v
Executing command (/Users/veilig/Sites/magento.magestead.clean): git describe --exact-match --tags
Reading /Users/veilig/.composer/composer.json
Loading config file /Users/veilig/.composer/config.json
Loading config file /Users/veilig/.composer/auth.json
Loading config file /Users/veilig/.composer/composer.json
Loading config file /Users/veilig/.composer/auth.json
Reading /Users/veilig/Sites/magento.magestead.clean/vendor/composer/installed.json
Reading /Users/veilig/.composer/vendor/composer/installed.json
Loading plugin AydinHassan\MagentoCoreComposerInstaller\CoreManager
Loading plugin MagentoHackathon\Composer\Magento\Plugin
you may want to add the packages.magento.com repository to composer.
add it with: composer.phar config -g repositories.magento composer https?://packages.magento.com
activate magento plugin
Running 1.1.1 (2016-05-17 12:25:44) with PHP 5.6.21 on Darwin / 15.4.0
remove all deployed modules
start magento module deploy via moduleManager
Start magento deploy for inchoo/php7
Start magento deploy for inchoo/php7
jump over deployLibraries as no Magento libraryPath is set
Autoloader patch to public/app/Mage.php was applied successfully
```

now without doing anything else, when I try to load my new VM in my browser, I get autoloading issues.

`Warning: include(PhpParser\Autoloader.php): failed to open stream: No such file or directory in /var/www/public/lib/Varien/Autoload.php on line 94`

If I don't boostrap and skip applying the autoload patch, everything loads correctly.  Have you encountered any issue like this?

**Comment by richdynamix** [comment]
Unfortunately I have seen this before. I removed the autoload patch due to this issue. You can see the commit here https://github.com/richdynamix/magestead/commit/743f2a8aae72da20bded774823b5246d099f4d74

It was working perfectly up to that day and I spent some time trying to get to the bottom of it with little success. I didn't want the issue delaying other work on the project so opted to remove it till later. If I recall correctly the issue disappeared for a while but re-appeared so I quickly removed it again. 

I am happy to receive any help from the wider community on solving this issue. It would be good to get solved for good.

**Comment by veilig2000** [comment]
I walked through to find out what is going on.  To fix the issue I put an ugly psr-4 hack into the root composer.json so the composer autoloader knows where to find the PhpParser autoloader file

```
    "autoload": {
        "psr-0": {
            "MageTest\\PhpSpec\\MagentoExtension": "src",
            "": [
                "public/app/code/local",
                "public/app/code/community",
                "public/app/code/core",
                "public/app",
                "public/lib",
                "features/bootstrap"
            ],
            "Mage": "public/app/code/core"
        },
        "psr-4": {
            "PhpParser\\": "vendor/nikic/php-parser/lib/PhpParser/"
        } 
    } 
```

The gist of the issue is, the composer autoloader gets loaded prior to calling the varien autoloader which enables it to find the lib/bootstrap.php file of the nikic/php-parser package due to its composer autoload config setup.
ie:

```
    "autoload": {
        "files": ["lib/bootstrap.php"]
    },
```

 But then inside that package they use their own autoloading which runs into a chicken or the egg race.  When they try to check if the `PhpParser/Autoload` class exists in the bootstrap.php file there are only two autoload functions registered at the time (executed in this order)
1. composer - which has no clue how to autoload those files b/c they use they register their own autoloader
2. varien - which also doesn't know 

So ultimately the autoload just falls through both functions registered and varien ends up failing.  Hence why my hack worked, b/c it allowed the composer autoloader to find the file/class w/ psr4.

Its an ugly fix, but I'm not sure you'll be able to do anything w/out the modification of that package.

**Comment by richdynamix** [comment]
I am not entirely convinced you can call this a hack. It does what it was intended, it setups up the correct autoloading. 

Since the patch itself is the real hack of Magento, which is also required to get composer libs into the site, I think this should be added to the main composer file.

Will update and push for the next release.

**Comment by richdynamix** [comment]
Fix has been applied and added to version 2.1.1
